lib/commands/network.js

```diff
@@ -42,6 +45,116 @@ commands.setNetworkConnection = async function (type) {
+  });
+};
+
+commands.toggleFlightMode = async function () {
+  log.info('Toggling flight mode');
+  let flightMode = !(await this.adb.isAirplaneModeOn());
+  await this.wrapBootstrapDisconnect(async () => {
+    await this.adb.setAirplaneMode(flightMode);
+  });
+  await this.wrapBootstrapDisconnect(async () => {
+    await this.adb.broadcastAirplaneMode(flightMode);
+  });
+};
+
+commands.setGeoLocation = async function (location) {
+  let cmd = `geo fix ${location.longitude} ${location.latitude}`;
+  return await this.adb.sendTelnetCommand(cmd);
+};
+
+commands.toggleLocationServices = async function(){
+  log.info('Toggling location services');
+  let api = await this.adb.getApiLevel();
+  
+  if (api > 15) {
+    let seq = [19, 19]; // up, up
+    if (api === 16) {
+      // This version of Android has a "parent" button in its action bar
+      seq.push(20); // down
+    } else if (api >= 19) {
+      // Newer versions of Android have the toggle in the Action bar
+      seq = [22, 22, 19]; // right, right, up
+      /*
+       * Once the Location services switch is OFF, it won't receive focus
+       * when going back to the Location Services settings screen unless we
+       * send a dummy keyevent (UP) *before* opening the settings screen
+       */
+      await this.adb.keyevent(19);
+    }
+    await this.toggleSetting('LOCATION_SOURCE_SETTINGS', seq);
+  } else {
+    // There's no global location services toggle on older Android versions
+    throw new errors.NotYetImplementedError();
+  }
+};
+
+helpers.toggleSetting = async function(setting, preKeySeq) {
+  /*
+   * preKeySeq is the keyevent sequence to send over ADB in order
+   * to position the cursor on the right option.
+   * By default it's [up, up, down] because we usually target the 1st item in
+   * the screen, and sometimes when opening settings activities the cursor is
+   * already positionned on the 1st item, but we can't know for sure
+   */
+  if (_.isNull(preKeySeq)) {
+    preKeySeq = [19, 19, 20]; // up, up, down
+  }
+
+  await this.openSettingsActivity(setting);
+  
+  for (let key of preKeySeq) {
+    await this.doKey(key);
+  }
+
+  let {appPackage, appActivity} = await this.adb.getFocusedPackageAndActivity();
+
+  /*
+   * Click and handle potential ADB disconnect that occurs on official
+   * emulator when the network connection is disabled
+   */
+  await this.wrapBootstrapDisconnect(async () => {
+    await this.doKey(23);
+  });
+
+  /*
+   * In one particular case (enable Location Services), a pop-up is
+   * displayed on some platforms so the user accepts or refuses that Google
+   * collects location data. So we wait for that pop-up to open, if it
+   * doesn't then proceed
+   */
+  try {
+    await this.adb.waitForNotActivity(appPackage, appActivity, 5000);
+    await this.doKey(22); // right
+    await this.doKey(23); // click
+    await this.adb.waitForNotActivity(appPackage, appActivity, 5000);
+  } catch(e) {
```

Comment:
a useful convention for this (ignoring a caught error) is:

```js
} catch (ign) {}
```

(to explicitly say we're ignoring the error)